src/tuner/algo/allreduce/ring.h

```diff
@@ -0,0 +1,289 @@
+ */
+#define NCCL_OFI_TUNER_RING_LL128_OVERHEAD_NUMERATOR    (15ULL)
+
+/**
+ * @brief Overhead denominator for NCCL_PROTO_LL128 protocol.
+ */
+#define NCCL_OFI_TUNER_RING_LL128_OVERHEAD_DENOMINATOR  (16ULL)
+
+/**
+ * @brief Overhead numerator for NCCL_PROTO_LL protocol.
+ */
+#define NCCL_OFI_TUNER_RING_LL_OVERHEAD_NUMERATOR       (1ULL)
+
+/**
+ * @brief Overhead denominator for NCCL_PROTO_LL protocol.
+ */
+#define NCCL_OFI_TUNER_RING_LL_OVERHEAD_DENOMINATOR     (2ULL)
+
+/**
+ * @brief Buffer size for NCCL_PROTO_LL128 protocol.
+ */
+#define NCCL_OFI_TUNER_NCCL_LL128_BUFFSIZE                                                     \
+	(NCCL_OFI_TUNER_NCCL_LL128_ELEMS_PER_THREAD * NCCL_OFI_TUNER_NCCL_LL128_MAX_NTHREADS * \
+	 NCCL_OFI_TUNER_NCCL_STEPS * sizeof(uint64_t))
+
+/**
+ * @brief Buffer size for NCCL_PROTO_LL protocol.
+ */
+#define NCCL_OFI_TUNER_NCCL_LL_BUFFSIZE                                                  \
+	(NCCL_OFI_TUNER_NCCL_LL_LINES_PER_THREAD * NCCL_OFI_TUNER_NCCL_LL_MAX_NTHREADS * \
+	 NCCL_OFI_TUNER_NCCL_STEPS * NCCL_OFI_TUNER_NCCL_SIZEOF_NCCL_LL_FIFOLINE)
+
+/**
+ * @brief Get the overhead numerator for a given protocol.
+ *
+ * @param proto The protocol (NCCL_PROTO_SIMPLE, NCCL_PROTO_LL, or NCCL_PROTO_LL128).
+ * @return The overhead numerator for the given protocol.
+ */
+static inline unsigned long long get_overhead_numerator(int proto)
+{
+	if (NCCL_PROTO_LL == proto) {
+		return NCCL_OFI_TUNER_RING_LL_OVERHEAD_NUMERATOR;
+	} else if (NCCL_PROTO_LL128 == proto) {
+		return NCCL_OFI_TUNER_RING_LL128_OVERHEAD_NUMERATOR;
+	} else if (NCCL_PROTO_SIMPLE == proto) {
+		return NCCL_OFI_TUNER_RING_SIMPLE_OVERHEAD_NUMERATOR;
+	}
+	assert(false);
+	__builtin_unreachable();
+}
+
+/**
+ * @brief Get the overhead denominator for a given protocol.
+ *
+ * @param proto The protocol (NCCL_PROTO_SIMPLE, NCCL_PROTO_LL, or NCCL_PROTO_LL128).
+ * @return The overhead denominator for the given protocol.
+ */
+static inline unsigned long long get_overhead_denominator(int proto)
+{
+	if (NCCL_PROTO_LL == proto) {
+		return NCCL_OFI_TUNER_RING_LL_OVERHEAD_DENOMINATOR;
+	} else if (NCCL_PROTO_LL128 == proto) {
+		return NCCL_OFI_TUNER_RING_LL128_OVERHEAD_DENOMINATOR;
+	} else if (NCCL_PROTO_SIMPLE == proto) {
+		return NCCL_OFI_TUNER_RING_SIMPLE_OVERHEAD_DENOMINATOR;
+	}
+	assert(false);
+	__builtin_unreachable();
+}
+
+/**
+ * @brief Get the buffer size for a given protocol.
+ *
+ * @param proto The protocol (NCCL_PROTO_SIMPLE, NCCL_PROTO_LL, or NCCL_PROTO_LL128).
+ * @return The buffer size for the given protocol.
+ */
+static inline unsigned long long get_buffsize(int proto,
+					      struct nccl_ofi_tuner_model_params const *params)
+{
+	if (NCCL_PROTO_LL == proto) {
+		return NCCL_OFI_TUNER_NCCL_LL_BUFFSIZE;
+	} else if (NCCL_PROTO_LL128 == proto) {
+		return NCCL_OFI_TUNER_NCCL_LL128_BUFFSIZE;
+	} else if (NCCL_PROTO_SIMPLE == proto) {
+		return params->nccl_buffsize;
+	}
+	assert(false);
```

Comment:
We should return either 0 or infinite cost here.  But should handle NCCL adding a new protocol in the future.

Reply:
one of the advantages of the header versioning approach is that we can control when this happens -- I'd hope that any tunerV2_as_V4 interface would handle this for us and not consult the external plugin against protocols that weren't defined at the time -- but a good call out. 

The problem is that there's not a great sentinel to return here that ensures callers will check it, and if things cancel-out it can be elided. https://godbolt.org/z/Ec8jr1vM6

Reply:
I would not expect NCCL to behave that nicely.  I think what you have is the worst possible option.  If you feel that strongly about not having a "obviously invalid", then please check and call abort() here.  Do not rely on asserts.

Reply:
Changed it into an abort call